### DISCUSSION TITLE
Failed use of recognize

### DISCUSSION THREAD

**Discussion by bawolk** [winnow-rs/winnow#448] [Q&A]
I need to parse an identifier string, which starts with an alphabetic character followed by none or more alphanumeric characters or an underscore.  An example would be "hello_123".   I tried
`
let mut parser = (
            alpha1,
            repeat(0.., alt(('_', alphanumeric1))),
        )
            .recognize();
        assert_eq!(parser.parse_peek("hello_123"), Ok(("", "hello_123")));`
But this does not compile.  What is the appropriate was to parse such an identifier?

**Comment by epage** [comment]
The first problem you likely ran into is `alt` requires each case to have the same data type.

You can find a working example of the code at https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=f15e9839c554a01c5ac95bd1441fde84

I would do this slightly differently though.  I would do something like
```rust
const ALPHA = ('a'..='z', 'A'..='Z');
const DIGIT = '0'..='9';
let mut parser = (one_of(ALPHA), take_while(0.., (ALPHA, DIGIT, '_))).recognize();
```
